### PR TITLE
Swift4 Remaster

### PR DESCRIPTION
**Pull request by lozhuf**
ShopGunSDK, remastered!

This is a total Swift 4 rewrite of the SDK (Only commits after Jan 04, 2018 need be considered)

Please have a look at the preliminary ReadMe, and it'd be great to hear your feedback on the public interface (esp. how the SDK is configured - let me know if it makes sense). Also, if you checkout the code you'll be able to visit the documentation site at `/docs/index.html`

### CODE REVIEW

**Comment by mortenbo** [comment]
Awesome! :D

**Comment by lozhuf** [comment]
Just some brainstorming about the configuration interface.... 

Currently it is done for all components through a central `ShopGun` namespace, and access to the configured component instances is also through that namespace.
```swift
ShopGun.configure(coreAPI: CoreAPI.Settings(...),
            eventsTracker: EventsTracker.Settings(...),
            dataStore: DataStore.Settings(…))

ShopGun.coreAPI.request(…)
ShopGun.eventsTracker.track(…)
```
But this means that the components (CoreAPI/EventsTracker/GraphAPI) cant be imported separately with subspecs in cocoapods (as the ShopGun class needs to know about all the other components).

One nice aspect of this design is that we can have a `ShopGun.configure()` method that reads from a single .plist and configures the different components based on the properties it reads.

So I've been considering possible alternatives:

#### 1) Remove the central namespace

So each component is configured separately in the AppDelegate.
```swift
DataStore.configure(DataStore.Settings(…))
CoreAPI.configure(CoreAPI.Settings(...))
EventsTracker.configure(EventsTracker.Settings(...))

CoreAPI.shared.request(…)
EventsTracker.shared.request(…)
```
Main con with this system is shared dependancies - both components need access to a pre-configured keychain manager, which would need to be initialized before either component is configured, leading to possible user error. Maybe it doesnt need to be a shared datastore? But then they would have to configure it separately for each component, which would be kind of annoying.

Also, I'm not a fan of the `.shared` interface (tho it is basically the same as `ShopGun.coreAPI`)

Benefit of this is that it'd be easy to re-configure an individual component while the app is running.

#### 2) Make the settings generic

So you configure with an array of `ConfigurationSettings` with the central `ShopGun` namespace, but  its the settings themselves that know how to do the setup.

```swift
ShopGun.configure(dataStore: DataStore.Settings(…), 
                             components: [CoreAPI.Settings(...), EventsTracker.Settings(...), ...])
ShopGun.coreAPI.request(…)
```

Main con if this is that it's a bit weird looking, its a bit strange to have the settings know how to configure the owner, and you could pass settings for the same component twice.

#### 3) ShopGun namespace but separate config methods

Use stand alone config methods for each component, in an extension on ShopGun, so it can be in separate subspecs.
```swift
ShopGun.configure(dataStore: DataStore.Settings(…))
ShopGun.configure(coreAPI: CoreAPI.Settings(…))
ShopGun.configure(eventsTracker: EventsTracker.Settings(…))

ShopGun.coreAPI.request(…)
ShopGun.eventsTracker.track(…)
```
Basically the same as **1)**, except that it doesnt have to use the `.shared` accessor.

Also same con as in **1)**... shared datastore needs to be configured first. Possible work around is to have a shared `ShopGun.dataStore` be accessed by all components that need it when they need it, rather than passing a dataStore instance into the component on init. This singleton pattern is not good for testing.
